src/OmniSharp.MSBuild/MSBuildEnvironment.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System;
+using System.IO;
+using Microsoft.Extensions.Logging;
+
+namespace OmniSharp.MSBuild
+{
+    public static class MSBuildEnvironment
+    {
+        private static bool s_isInitialized;
+        private static string s_msbuildFolder;
+
+        public static bool IsInitialized => s_isInitialized;
+
+        public static string MSBuildFolder
+        {
+            get
+            {
+                EnsureInitialized();
+                return s_msbuildFolder;
+            }
+        }
+
+        public static void Initialize(ILogger logger)
+        {
+            if (s_isInitialized)
+            {
+                throw new InvalidOperationException("MSBuildEnvironment is already initialized");
+            }
+
+            // If OmniSharp is running normally, MSBuild is located in an 'msbuild' folder beneath OmniSharp.exe.
+            var msbuildFolder = Path.Combine(AppContext.BaseDirectory, "msbuild");
+
+            if (!Directory.Exists(msbuildFolder))
+            {
+                // If the 'msbuild' folder does not exist beneath OmniSharp.exe, this is likely a development scenario,
+                // such as debugging or running unit tests. In that case, we use one of the .msbuild-* folders at the
+                // solution level.
+                msbuildFolder = FindMSBuildFolderFromSolution();
+            }
+
+            if (msbuildFolder == null || !Directory.Exists(msbuildFolder))
```

Comment:
`FindMSBuildFolderFromSolution()` already checks for directory internally so no need to recheck again, just `null` check would be enough